File: frontend/src/services/moduleService.ts
/**
 * Servicio para gestión de módulos
 * Comunicación con API backend para operaciones CRUD de módulos
 */
import axios, { AxiosResponse, AxiosRequestConfig } from 'axios';
import { 
  ModuleResponse, 
  ModuleCreate, 
  ModuleUpdate, 
  ModuleWithLessons, 
  CourseStructureResponse,
  LessonAssignment,
  CourseModulesResponse
} from '../types/module';
import { buildApiUrl, getAuthHeaders, API_ENDPOINTS } from '../config/api.config';

// Interface for API error responses - Following CLAUDE.md TypeScript best practices
interface APIError extends Error {
  response?: {
    status?: number;
    data?: {
      detail?: string;
      message?: string;
    };
  };
}

class ModuleService {
  private getAuthHeaders(): AxiosRequestConfig {
    return {
      headers: getAuthHeaders()
    };
  }

  /**
   * Crear un nuevo módulo para un curso
   */
  async createModule(courseId: string, moduleData: ModuleCreate): Promise<ModuleResponse> {
    const url = buildApiUrl(`${API_ENDPOINTS.COURSES}/${courseId}/modules`);
    console.log('🚀 [moduleService] Creating module for course:', courseId);
    console.log('📤 [moduleService] Request URL:', url);
    console.log('📤 [moduleService] Request payload:', JSON.stringify(moduleData, null, 2));
    console.log('📤 [moduleService] Request headers:', JSON.stringify(this.getAuthHeaders(), null, 2));

    try {
      const response: AxiosResponse<ModuleResponse> = await axios.post(
        url,
        moduleData,
        this.getAuthHeaders()
      );

      console.log('✅ [moduleService] Module created:', response.data.title);
      return response.data;
    } catch (error: any) {
      console.error('❌ [moduleService] Full error object:', error);
      console.error('❌ [moduleService] Error response:', error.response?.data);
      console.error('❌ [moduleService] Error status:', error.response?.status);
      console.error('❌ [moduleService] Error headers:', error.response?.headers);

      const apiError = error as APIError;
      const errorMessage = apiError.response?.data?.detail ||
                          apiError.response?.data?.message ||
                          apiError.message ||
                          'Error al crear módulo';

      throw new Error(errorMessage);
    }
  }

  /**
   * Obtener todos los módulos de un curso
   */
  async getCourseModules(courseId: string): Promise<ModuleResponse[]> {
    console.log('🔍 [moduleService] Getting modules for course:', courseId);
    
    try {
      const response: AxiosResponse<ModuleResponse[]> = await axios.get(
        buildApiUrl(`${API_ENDPOINTS.COURSES}/${courseId}/modules`)
      );
      
      console.log('✅ [moduleService] Modules fetched:', response.data.length, 'modules');
      return response.data;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error fetching modules:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al obtener módulos');
    }
  }

  /**
   * Obtener estructura completa del curso con módulos y lecciones
   */
  async getCourseStructure(courseId: string): Promise<CourseStructureResponse> {
    console.log('🏗️ [moduleService] Getting course structure for:', courseId);
    
    try {
      const response: AxiosResponse<CourseStructureResponse> = await axios.get(
        buildApiUrl(`${API_ENDPOINTS.COURSES}/${courseId}/structure`)
      );
      
      console.log('✅ [moduleService] Course structure fetched:', {
        modules: response.data.modules?.length || 0,
        total_lessons: response.data.total_lessons || 0
      });
      return response.data;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error fetching course structure:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al obtener estructura del curso');
    }
  }

  /**
   * Obtener un módulo específico por ID
   */
  async getModule(moduleId: string): Promise<ModuleResponse> {
    console.log('🔍 [moduleService] Getting module:', moduleId);
    
    try {
      const response: AxiosResponse<ModuleResponse> = await axios.get(
        buildApiUrl(`/modules/${moduleId}`)
      );
      
      console.log('✅ [moduleService] Module fetched:', response.data.title);
      return response.data;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error fetching module:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al obtener módulo');
    }
  }

  /**
   * Obtener un módulo con sus lecciones
   */
  async getModuleWithLessons(moduleId: string): Promise<ModuleWithLessons> {
    console.log('📚 [moduleService] Getting module with lessons:', moduleId);
    
    try {
      const response: AxiosResponse<ModuleWithLessons> = await axios.get(
        buildApiUrl(`/modules/${moduleId}/with-lessons`)
      );
      
      console.log('✅ [moduleService] Module with lessons fetched:', {
        title: response.data.title,
        lessons_count: response.data.lessons?.length || 0
      });
      return response.data;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error fetching module with lessons:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al obtener módulo con lecciones');
    }
  }

  /**
   * Actualizar un módulo
   */
  async updateModule(moduleId: string, moduleData: ModuleUpdate): Promise<ModuleResponse> {
    console.log('✏️ [moduleService] Updating module:', moduleId, moduleData);
    
    try {
      const response: AxiosResponse<ModuleResponse> = await axios.put(
        buildApiUrl(`/modules/${moduleId}`),
        moduleData,
        this.getAuthHeaders()
      );
      
      console.log('✅ [moduleService] Module updated:', response.data.title);
      return response.data;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error updating module:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al actualizar módulo');
    }
  }

  /**
   * Eliminar un módulo
   */
  async deleteModule(moduleId: string): Promise<void> {
    console.log('🗑️ [moduleService] Deleting module:', moduleId);
    
    try {
      await axios.delete(
        buildApiUrl(`/modules/${moduleId}`),
        this.getAuthHeaders()
      );
      
      console.log('✅ [moduleService] Module deleted successfully');
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error deleting module:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al eliminar módulo');
    }
  }

  /**
   * Reordenar módulos de un curso
   */
  async reorderModules(courseId: string, moduleOrders: Array<{ id: string; order: number }>): Promise<void> {
    console.log('🔄 [moduleService] Reordering modules for course:', courseId, moduleOrders);
    
    try {
      await axios.put(
        buildApiUrl(`${API_ENDPOINTS.COURSES}/${courseId}/modules/reorder`),
        moduleOrders,
        this.getAuthHeaders()
      );
      
      console.log('✅ [moduleService] Modules reordered successfully');
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error reordering modules:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al reordenar módulos');
    }
  }

  /**
   * Asignar una lección a un módulo
   */
  async assignLessonToModule(moduleId: string, lessonId: string): Promise<void> {
    console.log('🔗 [moduleService] Assigning lesson to module:', { moduleId, lessonId });
    
    try {
      await axios.post(
        buildApiUrl(`/modules/${moduleId}/lessons`),
        { lesson_id: lessonId } as LessonAssignment,
        this.getAuthHeaders()
      );
      
      console.log('✅ [moduleService] Lesson assigned to module successfully');
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error assigning lesson to module:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al asignar lección al módulo');
    }
  }

  /**
   * Remover una lección de un módulo
   */
  async removeLessonFromModule(moduleId: string, lessonId: string): Promise<void> {
    console.log('🔗 [moduleService] Removing lesson from module:', { moduleId, lessonId });
    
    try {
      await axios.delete(
        buildApiUrl(`/modules/${moduleId}/lessons/${lessonId}`),
        this.getAuthHeaders()
      );
      
      console.log('✅ [moduleService] Lesson removed from module successfully');
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [moduleService] Error removing lesson from module:', apiError);
      throw new Error(apiError.response?.data?.detail || 'Error al remover lección del módulo');
    }
  }

  // Métodos de utilidad para el frontend

  /**
   * Formatear duración en minutos a formato legible
   */
  formatDuration(minutes: number): string {
    if (minutes < 60) {
      return `${minutes} min`;
    }
    const hours = Math.floor(minutes / 60);
    const remainingMinutes = minutes % 60;
    if (remainingMinutes === 0) {
      return `${hours}h`;
    }
    return `${hours}h ${remainingMinutes}min`;
  }

  /**
   * Calcular progreso de módulo basado en lecciones completadas
   */
  calculateModuleProgress(module: ModuleWithLessons): number {
    if (module.lessons_count === 0) return 0;
    
    const completedLessons = (Array.isArray(module.lessons) ? module.lessons : []).filter(lesson => lesson.is_completed).length;
    return Math.round((completedLessons / module.lessons_count) * 100);
  }

  /**
   * Verificar si un módulo está bloqueado
   */
  isModuleLocked(module: ModuleWithLessons, previousModules: ModuleWithLessons[]): boolean {
    if (!module.unlock_previous) return false;
    
    const previousModule = previousModules.find(m => m.order === module.order - 1);
    if (!previousModule) return false;
    
    return this.calculateModuleProgress(previousModule) < 100;
  }

  /**
   * Obtener siguiente lección disponible en el curso
   */
  getNextLesson(courseStructure: CourseStructureResponse, currentLessonId?: string): {
    moduleId: string;
    lessonId: string;
    title: string;
  } | null {
    if (!currentLessonId) {
      // Devolver primera lección del primer módulo
      const firstModule = courseStructure.modules.find(m => m.order === 1);
      if (firstModule && firstModule.lessons.length > 0) {
        const firstLesson = firstModule.lessons.find(l => l.order === 1);
        if (firstLesson) {
          return {
            moduleId: firstModule.id,
            lessonId: firstLesson.id,
            title: firstLesson.title
          };
        }
      }
      return null;
    }

    // Encontrar lección actual y devolver la siguiente
    for (const module of courseStructure.modules) {
      const currentLessonIndex = module.lessons.findIndex(l => l.id === currentLessonId);
      
      if (currentLessonIndex !== -1) {
        // Si hay una siguiente lección en el mismo módulo
        if (currentLessonIndex < module.lessons.length - 1) {
          const nextLesson = module.lessons[currentLessonIndex + 1];
          return {
            moduleId: module.id,
            lessonId: nextLesson.id,
            title: nextLesson.title
          };
        }
        
        // Si no, buscar primera lección del siguiente módulo
        const nextModule = courseStructure.modules.find(m => m.order === module.order + 1);
        if (nextModule && nextModule.lessons.length > 0) {
          const firstLesson = nextModule.lessons.find(l => l.order === 1);
          if (firstLesson) {
            return {
              moduleId: nextModule.id,
              lessonId: firstLesson.id,
              title: firstLesson.title
            };
          }
        }
        
        return null; // No hay más lecciones
      }
    }
    
    return null;
  }
}

export default new ModuleService();

File: backend/app/application/services/module_service.py
"""
Servicio de Módulos
Lógica de negocio para gestión de módulos de cursos
"""
from datetime import datetime
from typing import List, Optional, Dict, Any
from app.domain.entities.module import Module
from app.domain.repositories.module_repository import ModuleRepository
from app.domain.repositories.course_repository import CourseRepository
from app.domain.repositories.lesson_repository import LessonRepository

class ModuleService:
    
    def __init__(
        self, 
        module_repository: ModuleRepository,
        course_repository: CourseRepository,
        lesson_repository: LessonRepository
    ):
        self.module_repository = module_repository
        self.course_repository = course_repository
        self.lesson_repository = lesson_repository
    
    async def create_module(self, course_id: str, module_data: Dict[str, Any]) -> Optional[Module]:
        """Crear un nuevo módulo en un curso"""
        # Verificar que el curso existe
        course = await self.course_repository.get_by_id(course_id)
        if not course:
            return None
        
        # Si no se especifica order, usar el siguiente disponible
        if "order" not in module_data:
            modules_count = await self.module_repository.count_by_course(course_id)
            module_data["order"] = modules_count + 1
        
        # Crear módulo
        module = Module(
            course_id=course_id,
            **module_data
        )
        
        return await self.module_repository.create(module)
    
    async def get_module_by_id(self, module_id: str) -> Optional[Module]:
        """Obtener módulo por ID"""
        return await self.module_repository.get_by_id(module_id)
    
    async def get_course_modules(self, course_id: str) -> List[Module]:
        """Obtener todos los módulos de un curso ordenados"""
        return await self.module_repository.get_by_course_id(course_id)
    
    async def get_module_with_lessons(self, module_id: str) -> Optional[Dict[str, Any]]:
        """Obtener módulo con información completa de sus lecciones"""
        module = await self.module_repository.get_by_id(module_id)
        if not module:
            return None
        
        # Obtener información de las lecciones
        lessons = []
        for lesson_id in module.lessons:
            lesson = await self.lesson_repository.get_by_id(lesson_id)
            if lesson:
                lessons.append(lesson)
        
        # Ordenar lecciones por order
        lessons.sort(key=lambda x: x.order)
        
        return {
            "module": module,
            "lessons": lessons,
            "lessons_count": len(lessons),
            "total_duration": sum(lesson.duration for lesson in lessons)
        }
    
    async def update_module(self, module_id: str, module_data: Dict[str, Any]) -> Optional[Module]:
        """Actualizar módulo"""
        # Verificar que el módulo existe
        existing_module = await self.module_repository.get_by_id(module_id)
        if not existing_module:
            return None
        
        # Si se cambia el order, verificar conflictos
        if "order" in module_data:
            new_order = module_data["order"]
            if new_order != existing_module.order:
                await self._handle_order_change(
                    existing_module.course_id, 
                    module_id, 
                    existing_module.order, 
                    new_order
                )
        
        return await self.module_repository.update(module_id, module_data)
    
    async def delete_module(self, module_id: str) -> bool:
        """Eliminar módulo y reorganizar orden"""
        # Obtener módulo antes de eliminar
        module = await self.module_repository.get_by_id(module_id)
        if not module:
            return False
        
        # Eliminar módulo
        deleted = await self.module_repository.delete(module_id)
        
        if deleted:
            # Reordenar módulos restantes
            await self._reorder_after_deletion(module.course_id, module.order)
        
        return deleted
    
    async def reorder_course_modules(self, course_id: str, module_orders: List[Dict[str, int]]) -> bool:
        """Reordenar todos los módulos de un curso"""
        return await self.module_repository.reorder_modules(course_id, module_orders)
    
    async def add_lesson_to_module(self, module_id: str, lesson_id: str) -> bool:
        """Agregar lección a módulo"""
        # Verificar que tanto módulo como lección existen
        module = await self.module_repository.get_by_id(module_id)
        lesson = await self.lesson_repository.get_by_id(lesson_id)
        
        if not module or not lesson:
            return False
        
        # Verificar que la lección pertenece al mismo curso
        if lesson.course_id != module.course_id:
            return False
        
        return await self.module_repository.add_lesson_to_module(module_id, lesson_id)
    
    async def remove_lesson_from_module(self, module_id: str, lesson_id: str) -> bool:
        """Remover lección de módulo"""
        return await self.module_repository.remove_lesson_from_module(module_id, lesson_id)
    
    async def get_course_structure(self, course_id: str) -> Dict[str, Any]:
        """Obtener estructura completa del curso: módulos con sus lecciones"""
        modules = await self.get_course_modules(course_id)
        
        course_structure = {
            "course_id": course_id,
            "modules": [],
            "total_modules": len(modules),
            "total_lessons": 0,
            "total_duration": 0
        }
        
        for module in modules:
            module_data = await self.get_module_with_lessons(module.id)
            if module_data:
                # ✅ FIX: Aplanar estructura para coincidir con ModuleWithLessons
                module_dict = module_data["module"].dict()
                module_with_lessons = {
                    **module_dict,
                    "lessons": [lesson.dict() for lesson in module_data["lessons"]],
                    "lessons_count": module_data["lessons_count"],
                    "total_duration": module_data["total_duration"]
                }
                course_structure["modules"].append(module_with_lessons)
                course_structure["total_lessons"] += module_data["lessons_count"]
                course_structure["total_duration"] += module_data["total_duration"]
        
        return course_structure
    
    # Métodos privados para manejo interno
    
    async def _handle_order_change(self, course_id: str, module_id: str, old_order: int, new_order: int):
        """Manejar cambio de orden de módulo"""
        if old_order == new_order:
            return
        
        modules = await self.module_repository.get_by_course_id(course_id)
        
        # Reorganizar órdenes
        for module in modules:
            if module.id == module_id:
                continue
            
            current_order = module.order
            
            # Lógica de reordenamiento
            if old_order < new_order:
                # Mover hacia abajo: los módulos entre old_order y new_order suben
                if old_order < current_order <= new_order:
                    await self.module_repository.update(module.id, {"order": current_order - 1})
            else:
                # Mover hacia arriba: los módulos entre new_order y old_order bajan
                if new_order <= current_order < old_order:
                    await self.module_repository.update(module.id, {"order": current_order + 1})
    
    async def _reorder_after_deletion(self, course_id: str, deleted_order: int):
        """Reordenar módulos después de eliminación"""
        modules = await self.module_repository.get_by_course_id(course_id)
        
        # Bajar el orden de todos los módulos que estaban después del eliminado
        for module in modules:
            if module.order > deleted_order:
                await self.module_repository.update(module.id, {"order": module.order - 1})
